### PR TITLE
Adds CSP guide

### PR DESCRIPTION
**Pull request by el-mapache**
Adds content security policy information to the front end guide.

* Adds a new `Security` tab in navigation
* CSP section contains an overview, some usage suggestions, and links to more in depth reading and guides.

### CODE REVIEW

**Comment by toolness** [comment]
Nice!

CALC used to have CSP enabled, until we found out, about a year ago, that New Relic's browser-side functionality wasn't compatible with it at all.  This was a huge bummer... hopefully they have made it more compatible over the past year!

**Comment by el-mapache** [comment]
@toolness I believe you can include NR provided you pass a nonce attribute to the inline script tag! We do this on login and New Relic loads fine.

**Comment by toolness** [comment]
Hmm, so it's been about a year since I dealt with it, but I *think* that our New Relic script was being injected into the page on the server-side by the newrelic Python package, so we didn't actually have any control over the `<script>` tag itself, and it seems [folks on NR's support forum are still frustrated by this](https://discuss.newrelic.com/t/content-security-policy-and-browser-injection/2629/53).

In any case, it might be worth warning readers that some third-party services and/or libraries might not be very compatible with CSP (for example, I think Django's admin UI wasn't CSP compatible until fairly recently), so folks might want to do a bit of research before implementing it on their sites.

**Comment by el-mapache** [comment]
Ah yeah, thats a great point. I'll add that to the 'caveats' section!

File: pages/security/csp.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+## Caveats
+A major caveat is that CSP, by default, prevents developers from adding inline `<style>` and `<script>` tags, and from specifying JavaScript behavior or CSS `style` attributes directly on DOM elements. This isn't a bad thing, as current best practices in front-end development discourage the use of these patterns.
+
+However, there are scenarios in which you may want to use an inline content tag, such as including a Google Analytics initialization script. For these cases, CSP allows the user to supply a `nonce` attribute to the inline script: `<script nonce=GHDGsfsd83hfdfFD3>...javascript here...</script>`. Then, in your policy, `script-src 'nonce-GHDGsfsd83hfdfFD3'`. Nonces must be regenerated on each request. Additionally, CSP 2.0 allows you to generate a SHA hash of the script content, and pass that to the `srcipt-src` directive as `'sha-{content-sha}'`. You don't need to use a `nonce` attribute if your content is hashed.
+
+Another potential issue is third-party libraries that automatically inject JavaScript and CSS into your HTML. If your project utilizes a library that does this, the only solution is
```
Comment (el-mapache) [inline pages/security/csp.md]: This isn't strictly true, technically one could make a SHA256 hash of the contents of the injected javascript, and whitelist that. I didn't mention it because it seems really hacky and error-prone, as the hash would have to be recomputed every time the script changed.

That being said it is a valid method, and someone might decide it's an acceptable tradeoff, so maybe it's worth mentioning?

File: pages/security/csp.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+title: Content Security Policy
+permalink: /security/content-security-policy
+layout: docs
+sidenav: security
+---
+
+# Content Security Policy
+
+## What is Content Security Policy?
+Content Security Policy (referred to as **CSP** in the rest of this guide) is a security measure designed to mitigate the likelihood of Cross-Site Scripting (XXS) attacks and data injection.
```
Comment (xtine) [inline pages/security/csp.md]: Should be "XSS?"

Comment (el-mapache) [reply pages/security/csp.md]: Yes absolutely! Good catch.

File: pages/security/csp.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+`<meta http-equiv="Content-Security-Policy" content="default-src 'self'; img-src https://*; child-src 'none';">`
+
+You're done!
+
+### Server-side Implementation
+Similarly, to include CSP on the server, simply return a `Content-Security-Policy: {my-policy-string}` HTTP header in your web responses.
+
+<br>
+
+Although a basic CSP policy is relatively easy to write, you might find it easier to use a server-side library to define your policies declaratively. Two of well-known and actively maintained libraries are [secureheaders](https://github.com/twitter/secureheaders), a **ruby** gem written and maintained by Twitter, and the [helmetjs](https://github.com/helmetjs/helmet) middleware for **express.js**.
```
Comment (toolness) [inline pages/security/csp.md]: A lot of 18F projects use Django, and [django-cors-headers](https://github.com/ottoyiu/django-cors-headers) is the well-maintained, go-to solution for CORS on it. You're welcome to mention it too if you want, no worries if not!

Comment (toolness) [reply pages/security/csp.md]: Oh geez I just confused CORS with CSP. 😣 

The go-to CSP package is generally [django-csp](https://github.com/mozilla/django-csp).

File: pages/security/csp.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+## Caveats
+A major caveat is that CSP, by default, prevents developers from adding inline `<style>` and `<script>` tags, and from specifying JavaScript behavior or CSS `style` attributes directly on DOM elements. This isn't a bad thing, as current best practices in front-end development discourage the use of these patterns.
+
+However, there are scenarios in which you may want to use an inline content tag, such as including a Google Analytics initialization script. For these cases, CSP allows the user to supply a `nonce` attribute to the inline script: `<script nonce=GHDGsfsd83hfdfFD3>...javascript here...</script>`. Then, in your policy, `script-src 'nonce-GHDGsfsd83hfdfFD3'`. Nonces must be regenerated on each request. Additionally, CSP 2.0 allows you to generate a SHA hash of the script content, and pass that to the `srcipt-src` directive as `'sha-{content-sha}'`. You don't need to use a `nonce` attribute if your content is hashed.
```
Comment (toolness) [inline pages/security/csp.md]: It might also be worth mentioning backwards-[in]compatibility information about CSP here, e.g. what happens when a CSP 1.0-complaint browser encounters CSP 2.0 tags. My initial concern with this with CALC was that I didn't want the site to break on sites that were only 1.0 compliant, e.g. because they noticed the CSP headers and enabled CSP, but didn't understand the 2.0-style SHA hashes and ignored them, thus not allowing any scripts to load.  I found surprisingly little information on this after searching, and ultimately found (either through reading the spec or trial and error testing, I forget which) that 1.0-only browsers essentially disable CSP as soon as they encounter something in the CSP headers that they don't understand. This means that 1.0-only browsers won't break on a CSP 2.0 site--they'll just behave the same way CSP-noncompliant browsers behave.

It might be worth mentioning this for folks who were in the same confused state as me, though I understand it's a mouthful to explain. I'm happy to submit a separate PR after this lands, if you prefer.

File: pages/security/csp.md (hunk 1 above)
Comment (toolness) [reply pages/security/csp.md]: I'm fine with mentioning it with caveats!

File: pages/security/csp.md (hunk 4 above)
Comment (el-mapache) [reply pages/security/csp.md]: I do mention in the introduction that browsers will ignore directives that aren't recognized, but I can't find references to the entire policy being ignored.

As far as I can tell, only unsupported directives are ignored, and the rest are parsed normally. CSP 2.0 is supposed to be backwards compatible with 1.0, although I would totally believe there are cases where this doesn't hold!

I'm also totally down for any suggestions to add more visibility to this aspect of CSP.

Comment (toolness) [reply pages/security/csp.md]: Hmm, I could re-investigate if you want.  The main problem is that because simply activating CSP by default disables _everything_ and individual directives whitelist individual kinds of things, if a critical resource is made accessible only by a relatively new directive, that means that your site will break on browsers that support CSP but not that new directive.

I can try finding evidence of this being the rationale for why CSP doesn't follow the typical progressive enhancement strategy that most browser APIs follow. I really do wish that it were documented better, because it really scared me when I first started using CSP 2.

Comment (el-mapache) [reply pages/security/csp.md]: Ah ok, I _think_ that in the case you are describing, you would want a policy that at a minimum looks like:

```
"default-src 'unsafe-inline'; script-src 'nonce-${base64}'"
```

Additionally, the user could implement the directive like this:

```
"script-src 'unsafe-inline' https: 'nonce-${base64}'"
```
In each of these cases, browsers that don't support CSP 2.0, will fall back to the `unsafe-inline` directive, allowing inline scripts to execute. That does disable a key feature of CSP, but if the browser doesn't support CSP 2.0, and your site needs to make use of inline scripts, you would need to use `unsafe-inline` regardless.

Comment (toolness) [reply pages/security/csp.md]: Ohhh YES, that's it!!!! Totally. And it took me _so_ long to figure that out because it's not documented very well anywhere, and as you can see I've already forgotten what I had to do. It's also very easy to overlook, I think, since the above policy minus `unsafe-inline` would still work fine on CSP 2.0-compliant browsers and non-CSP browsers, but break utterly on CSP 1.0 browsers.

I think it would be really helpful to mention this in the guide. I'm happy to submit a PR after this lands, or if you want to, you can add it too, whatever works!

Comment (el-mapache) [reply pages/security/csp.md]: Oh good! I was afraid I was missing some context. You're right, this behavior is definitely not intuitive. I'm not 100% certain where to include this in the guide, (maybe it's too important to be in caveats?), but I agree we should! 

If there is a good spot for it that makes sense to you (as a consumer of the guide), please do submit a PR! Otherwise, I can take a stab at figuring out where the info should live.

**Comment by el-mapache** [comment]
just bumping this to see if its good to merge for now, or if there are further changes needed before we push it live! thanks!

**Comment by jeremiak** [comment]
I'd vote for merging because it looks good and I want to provide a link to this in some guidance I'm giving to the FBI.

File: pages/security/csp.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+It can be enabled on either client- or server-side, and is unobtrusive in unsupported browsers — that is, browsers that don't support CSP will ignore its directives and load the page as normal.
+
+CSP is supported by the current versions of **all modern desktop browsers**: Safari, Chrome, Firefox, and IE Edge. It is also supported in recent versions iOS Safari and Chrome for Android.
+
+Unfortunately, support for numbered versions of IE is essentially zero, with no support for IE < 10 and only two directives supported in IE 10+.
+
+All of the above supported browsers support CSP 1.0, with 100% support for 2.0 in Webkit/Blink browsers, and partial, but very good, support in Firefox. CSP 2.0 has all the features of 1.0, with several additional directives and support for inline `<script>` and `<style>` tags (see [Caveats](#caveats)).
+
+## Usage
+CSP is straightforward to implement, and supports providing a policy via HTTP header or `<meta>` tag.
```
Comment (hbillings) [inline pages/security/csp.md]: I think it might be helpful to reword this as, "...supports providing a policy either server-side (HTTP header) or client-side (`<meta>` tag). We could link down the page to those two sections also.

Can we use the example on line 84, from the client-side implementation section, to demonstrate what the meta tag looks like? I feel like that would be particularly helpful to folks who are new to this.

File: pages/security/csp.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+  child-src https://youtu.be : Allow embedded content
+    from YouTube over HTTPS
+
+  object-src none : Prevent Flash content from loading
+```
+
+The `default-src` directive should always be defined! This directive acts as a fallback for all other `*-src` directives that are not defined within in the policy or are unsupported.
+
+A full list of available directives and keywords is outside the scope of this guide, but is available in the excellent Google CSP guide linked in the [Further Reading](#further-reading) section.
```
Comment (hbillings) [inline pages/security/csp.md]: I would move this up to line 27, right after linking to the cheatsheet.

File: pages/security/csp.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+Another potential issue is third-party libraries that automatically inject JavaScript and CSS into your HTML. If your project utilizes a library that does this, the only guaranteed solution is
+to use the `unsafe-inline` value when setting the `script-src` directive; this obviously defeats the purpose of having a CSP for your JavaScript.
+
+If you must load external scripts inline and are not allowed to use the `unsafe-inline` keyword, you could also make a SHA hash of the script being included, and whitelist that in your content security policy.
+
+This technique will allow those scripts to load, with the following caveats:
+
+* Each time the third-party script changes, a new hash will have to be computed, and your policy's `script-src` whitelist will need to be updated
+* Inlining the SHAs of multiple scripts adds bloat to the policy, and the number of bytes needed to transmit the header to the browser.
```
Comment (hbillings) [inline pages/security/csp.md]: There's a word missing in the second half. Should it be "and increases the number of bytes"?

**Review by hbillings — DISMISSED** [review]
I have a few structural suggestions/questions, but otherwise this looks awesome! Thanks so much!

**Review by hbillings — DISMISSED** [review]

**Review by hbillings — APPROVED** [review]
